### PR TITLE
Decoder Attention CUDA Op

### PR DESCRIPTION
**Pull request by wangyems**
**Description**: Describe your changes.

This PR introduces the decoder attention cuda kernel. the interface definition can be found [here](https://github.com/microsoft/onnxruntime/blob/wangye/decoder_attn_pr/onnxruntime/core/graph/contrib_ops/contrib_defs.cc#L729)  

It can be considered as superset of existing attention kernel with the additional support of cross attention algorithm, transposed input, keys_padding_mask and flexible usage of key/value cache. This kernel also leverages on some existing shared cuda implementations from attention cuda kernel and add some new cuda kernels.

The entry point of this kernel is [here](https://github.com/microsoft/onnxruntime/blob/wangye/decoder_attn_pr/onnxruntime/contrib_ops/cuda/bert/decoder_attention.cc#L172)
 
The unit test can be found [here](https://github.com/microsoft/onnxruntime/blob/wangye/decoder_attn_pr/onnxruntime/test/contrib_ops/decoder_attention_op_test.cc)

The parity test can be found [here](https://github.com/microsoft/onnxruntime/blob/wangye/decoder_attn_pr/onnxruntime/test/python/transformers/test_parity_decoder_attention.py)

**Motivation and Context**
- Why is this change required? What problem does it solve?
improve the perf of a first party model

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Review by hanbitmyths — COMMENTED** [review]
Can you add a description to help reviewers, who doesn't know a background, understand what this PR is for before going through all code changes?

**Comment by wangyems** [comment]
> Can you add a description to help reviewers, who doesn't know a background, understand what this PR is for before going through all code changes?

sure

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu [hunk 1]
```diff
@@ -189,18 +194,246 @@ bool LaunchAttentionKernel(
...
+  const size_t element_size,
+  int batch_size,
+  int sequence_length,
+  int kv_sequence_length,
+  int num_heads,
+  int head_size,
+  bool static_kv,
+  bool use_past,
+  bool has_layer_state,
+  bool has_key_padding_mask,
```
Comment (hanbitmyths) [inline onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:224]: these are all const?

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu [hunk 2]
```diff
@@ -189,18 +194,246 @@ bool LaunchAttentionKernel(
...
```
Comment (hanbitmyths) [inline onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:369]: nit: const

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.h [hunk 3]
```diff
@@ -39,21 +39,71 @@ bool LaunchAttentionKernel(
...
+    const size_t element_size,                    // Element size of input tensor
+    int batch_size,                               // Batch size (B)
+    int sequence_length,                          // Sequence length (S)
+    int kv_sequence_length,                       // Key/Value/Cache sequence length
+    int num_heads,                                // Number of attention heads (N)
+    int head_size,                                // Hidden layer size per head (H)
+    bool static_kv,                               // Whether cross attention or not
+    bool use_past,                                // Whether use cache or not
+    bool has_layer_state,                         // Whether output cache or not
+    bool has_key_padding_mask,                    // Whether use key_padding_mask or not
```
Comment (hanbitmyths) [inline onnxruntime/contrib_ops/cuda/bert/attention_impl.h:55]: nit: const

File: onnxruntime/contrib_ops/cuda/bert/attention_transpose.cu [hunk 4]
```diff
@@ -156,9 +173,17 @@ __global__ void TransposeQKV(const int H, const T* input, T* output) {
...
   const int batch_size = gridDim.y;
+  const int chunk_num = gridDim.z;
   const int NH = num_heads * H;
   const int NHS = NH * sequence_length;
-  const int in_offset = n * H + m * NH + s * 3 * NH + b * NHS * 3;
+
+  int in_offset = 0;
+  if (reversed_bs) {
+    const int BNH = NH * batch_size;
+    in_offset = n * H + m * NH + b * chunk_num * NH + s * BNH * chunk_num;
```
Comment (hanbitmyths) [inline onnxruntime/contrib_ops/cuda/bert/attention_transpose.cu:183]: (m + b * chunk_num) * NH

File: onnxruntime/contrib_ops/cuda/bert/attention_transpose.cu [hunk 5]
```diff
@@ -156,9 +173,17 @@ __global__ void TransposeQKV(const int H, const T* input, T* output) {
...
+  } else {
+    in_offset = n * H + m * NH + s * chunk_num * NH + b * NHS * chunk_num;
```
Comment (hanbitmyths) [inline onnxruntime/contrib_ops/cuda/bert/attention_transpose.cu:185]: n * H + (m + s * chunk_num) * NH + b * NHS * chunk_num

File: onnxruntime/contrib_ops/cuda/bert/attention_transpose.cu [hunk 6]
```diff
@@ -184,9 +210,16 @@ __global__ void TransposeQKVLarge(const int H, const T* input, T* output) {
...
+  const int chunk_num = gridDim.z;
   const int NH = num_heads * H;
   const int NHS = NH * sequence_length;
-  const int in_offset = n * H + m * NH + s * 3 * NH + b * NHS * 3;
+  int in_offset = 0;
+  if (reversed_bs) {
+    const int BNH = NH * batch_size;
+    in_offset = n * H + m * NH + b * chunk_num * NH + s * BNH * chunk_num;
+  } else {
+    in_offset = n * H + m * NH + s * chunk_num * NH + b * NHS * chunk_num;
```
Comment (hanbitmyths) [inline onnxruntime/contrib_ops/cuda/bert/attention_transpose.cu:221]: same here

File: onnxruntime/contrib_ops/cuda/bert/decoder_attention.cc [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+  //k, v sequence after gemm
+  int kv_sequence_length = 0;
+
+  // Generate q, k, v w/o cache
+  // query input: (S, B, h1)
+  // key input: (S', B, h1)
+  // weight: (h1, h2)
+  // h = N*H
+  cublasHandle_t cublas = CublasHandle();
+  cublasSetStream(cublas, stream);
```
Comment (hanbitmyths) [inline onnxruntime/contrib_ops/cuda/bert/decoder_attention.cc:224]: check if successful

File: onnxruntime/contrib_ops/cuda/bert/decoder_attention.cc [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+          sequence_length,
+          kv_sequence_length,
+          num_heads_,
+          head_size,
+          static_kv_,
+          use_past_,
+          has_layer_state_,
+          has_key_padding_mask_,
+          nullptr == gemm_query_buffer_p? nullptr : reinterpret_cast<const CudaT*>(gemm_query_buffer_p.get()),
+          nullptr == gemm_kv_buffer_p? nullptr : reinterpret_cast<const CudaT*>(gemm_kv_buffer_p.get()),
```
Comment (hanbitmyths) [inline onnxruntime/contrib_ops/cuda/bert/decoder_attention.cc:368]: nit: space before '?'

File: onnxruntime/contrib_ops/cuda/bert/attention_transpose.cu [hunk 9]
```diff
@@ -180,9 +180,9 @@ __global__ void TransposeQKV(const int H, const bool reversed_bs, const T* input
   int in_offset = 0;
   if (reversed_bs) {
     const int BNH = NH * batch_size;
-    in_offset = n * H + m * NH + b * chunk_num * NH + s * BNH * chunk_num;
+    in_offset = n * H + (m + b * chunk_num) * NH + s * BNH * chunk_num;
   } else {
-    in_offset = n * H + m * NH + s * chunk_num * NH + b * NHS * chunk_num;
+    in_offset = n * H +  (m + s * chunk_num) * NH + b * NHS * chunk_num;
```
Comment (hanbitmyths) [inline onnxruntime/contrib_ops/cuda/bert/attention_transpose.cu:185]: nit: extra space here

File: onnxruntime/contrib_ops/cuda/bert/attention_impl.cu [hunk 10]
```diff
@@ -189,18 +193,246 @@ bool LaunchAttentionKernel(
...
+  const bool* key_padding_mask,
+  const T* key_cache,
+  const T* value_cache,
+  T* qkv_buffer,
+  T* workspace_buffer,
+  T* output,
+  T* new_key_cache,
+  T* new_value_cache)
+{
+  const int max_threads_per_block(prop.maxThreadsPerBlock);
```
Comment (hanbitmyths) [inline onnxruntime/contrib_ops/cuda/bert/attention_impl.cu:235]: nit: use the same initializer as others here even though there's no difference for int

**Review by hanbitmyths — APPROVED** [review]

**Pull request merged**